valohai_yaml/pipelines/conversion.py

```diff
@@ -14,6 +14,16 @@
 ConvertedObject = Dict[str, Any]
 
 
+class VariantExpression(TypedDict):
+    """Variant expression template."""
+
+    style: str
+    rules: Dict[str, Any]
+
+
+ExpressionValue = Union[str, int, bool, float, VariantExpression]
```

Comment:
How would this `VariantExpression` dictionary affect they way roi and/or `valohai-utils` parsing the value now? Btw, should we call it `ExpressionValueVariant`?

Reply:
Backend accepts variant expressions as parameter values, and it is the only way it accepts list parameters. It works for all parameter value types though.

Valohai-utils is not involved in this process at all. We pass a list in structured format to backend, and backend receives a list in structured format. If you pass a list parameter as a command line parameter and parse command line parameters in a Python script using `valohai-utils`, it must use comma separated values as that is the only format `valohai-utils` accepts. Other scripts or programs can accept different parameter passing styles, which is why we support custom multiple-separators.

You can also use file based configuration to read the structured list parameter as JSON (`valohai/config/parameters.json`), which skips the whole parsing step. This is what `valohai-utils` automatically does for parameters that haven't been overridden as command line parameters.